pkg/docker/config/config.go

```diff
@@ -37,111 +75,247 @@ var (
+			return err
+		}
+	}
+
+	if credHelper, exists := s.Config.CredHelpers[creds.ServerURL]; exists {
+		helperName := fmt.Sprintf("docker-credential-%s", credHelper)
+		p := helperclient.NewShellProgramFunc(helperName)
+		return helperclient.Store(p, creds)
+	}
+
+	newCreds := Auth{
+		Username: creds.Username,
+		Secret:   creds.Secret,
+	}
+	if newCreds != s.Config.Auths[creds.ServerURL] {
+		s.Config.Auths[creds.ServerURL] = newCreds
+		return s.save()
+	}
+
+	return nil
+}
+
+// Get retrieves credentials from the store.  It returns the username
+// and secret as strings.
+func (s *AuthStore) Get(serverURL string) (string, string, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.Config == nil {
+		err := s.load()
+		if err != nil {
+			return "", "", err
 		}
+	}
 
-		creds := base64.StdEncoding.EncodeToString([]byte(username + ":" + password))
-		newCreds := dockerAuthConfig{Auth: creds}
-		auths.AuthConfigs[registry] = newCreds
-		return true, nil
+	// First try cred helpers. They should always be normalized.
+	if credHelper, exists := s.Config.CredHelpers[serverURL]; exists {
+		helperName := fmt.Sprintf("docker-credential-%s", credHelper)
+		p := helperclient.NewShellProgramFunc(helperName)
+		creds, err := helperclient.Get(p, serverURL)
+		if err != nil {
+			return "", "", err
+		}
+		return creds.Username, creds.Secret, nil
+	}
+
+	// I'm feeling lucky
+	if val, exists := s.Config.Auths[serverURL]; exists {
+		return val.Username, val.Secret, nil
+	}
+
+	// bad luck; let's normalize the entries first
+	serverURL = normalizeRegistry(serverURL)
+	normalizedAuths := map[string]Auth{}
+	for k, v := range s.Config.Auths {
+		normalizedAuths[normalizeRegistry(k)] = v
+	}
+	if val, exists := normalizedAuths[serverURL]; exists {
+		return val.Username, val.Secret, nil
+	}
+	return "", "", credentials.NewErrCredentialsNotFound()
+}
+
+// Delete removes credentials from the store.
+func (s *AuthStore) Delete(serverURL string) error {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.Config == nil {
+		err := s.load()
+		if err != nil {
+			if credentials.IsErrCredentialsNotFound(err) {
+				return nil
+			}
+			return err
+		}
+	}
+
+	// First try cred helpers.
+	if credHelper, exists := s.Config.CredHelpers[serverURL]; exists {
+		helperName := fmt.Sprintf("docker-credential-%s", credHelper)
+		p := helperclient.NewShellProgramFunc(helperName)
+		return helperclient.Erase(p, serverURL)
+	}
+
+	changed := false
+	if _, ok := s.Config.Auths[serverURL]; ok {
+		delete(s.Config.Auths, serverURL)
+		changed = true
+	} else if _, ok := s.Config.Auths[normalizeRegistry(serverURL)]; ok {
+		delete(s.Config.Auths, normalizeRegistry(serverURL))
+		changed = true
+	}
+	if changed {
+		return s.save()
+	}
+
+	return nil
+}
+
+// SetAuthentication stores the username and password in the auth.json file
+//
+// Deprecated: Use an AuthStore.
+func SetAuthentication(sys *types.SystemContext, registry, username, password string) error {
```

Comment:
If `AuthStore` is supposed to be a replacement, it needs to support `SystemContext`.

Reply:
> If `AuthStore` is supposed to be a replacement, it needs to support `SystemContext`.

I don't think so.  The Docker client needs a way to get credentials for a given authority, and that's what the new `GetAuth` property provides.  Users will need a way to create that getter, and that's what the two new `AuthStore` structures provide.  The old `SystemContext` properties either configure a single credential (which can be replaced easily with the new `single.AuthStore`) or the path lookup for the Docker-config `AuthStore` (which can be replaced by creating your own `config.AuthStore` with `Path` set however they like).  The rest of `SystemContext` seems unrelated to managing authentication, right?